### PR TITLE
Limit number of concurrent background partition optimize

### LINKED ISSUE #53303: Consider adding a separated task for historical merges
**Use case**

The merged algorithm in ClickHouse was designed to favour smaller and more recent parts. This is obvious because we need to minimise number of parts.

However, there's a case when merging old parts also important. For example, when we have a `ReplacingMergeTree` and the `SELECT ... FINAL` spans across several partitions. Thanks to `do_not_merge_across_partitions_select_final`, if a partition only has 1 part, we don't need to apply `FINAL` logic on that partition, which significantly improve the query performance. Therefore we want that at some point, the "old" partition (last week, last month, etc...) will only contain 1 part.

**What we have now**

I think the use case has been raised in #35836, and there's already a solution with `min_age_to_force_merge_seconds` in #42423. With `min_age_to_force_merge_seconds`, once the parts stay long enough,  they will be allowed to merged without respecting the criteria (total size, age...).

However, this setting can lead to merge starving for recent parts, because: historical merges are usually big (not to say super big), and number of concurrent merges are limited.

```
-- Future merge tasks, mixing between small (recent parts) merges and big (old parts) merges
@@@@@@@@@
@
@@@@@@@@@@@
@
@@
@@@@@@@@@@@@@@
@@
-- Current merge tasks queue - running in round robin manner for each @
@
@@
@@@@
@@@@@
@@@@@@
@@@@@@
@@@@@@@@@@

-- Current merge tasks at some points
@@@@@@@@
@@@@@@@@@
@@@@@@@@@@@
@@@@@@
@@@@@@
@@@@@@
@@@@@@@
``` 

If we change `background_merges_mutations_scheduling_policy` to `shortest_task_first`, then we face the same old issue again: small merges are alway preferred, while big merges can stuck in merge queue forever (ever if they're scheduled).

**Describe the solution you'd like**

Each table having a separated task to optimize old partitions. The task will get thread from common schedule pool, and it will scan from oldest partition -> most recent partition to find which partition it can merge to single part. Whether to active this task or not will be controlled by a table setting. Merging old partition will takes some times, but eventually we will reach a point where every old partitions only have 1 part.

I think this is the original idea implemented in #35836, but then changed to current solution.

I don't know if there's a better solution, appreciate any comments [THANKS]!

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Limit number of concurrent background partition optimize merges

When enable `min_age_to_force_merge_seconds` and `min_age_to_force_merge_on_partition_only`, some merges will be generate to optimize an entire partition into a single part. This is similar to running `OPTIMIZE FINAL` on that partition, but using a thread from regular merge/mutate pool.

In some conditions, the number of "background partition optimize" merges becomes dominant and starves regular merges, which are more important.

This PR fix the issue by introduce a new merge tree setting, `number_of_free_entries_in_pool_to_execute_optimize_entire_partition` with default value = 20. With this setting, "background partition optimize" merges will only be scheduled if we have enough free entries in merge pool. 

This is the same way CH favouring merges over mutations with `number_of_free_entries_in_pool_to_execute_mutation`.

Close #53303, close #47420

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by canhld94** [comment]
Hi @tavplubix wdyt about this PR?

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit cd535bea7d85bdb0a763d817ae4920ead1d08f83 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53405/cd535bea7d85bdb0a763d817ae4920ead1d08f83/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by canhld94** [comment]
Segmentation fault in tsan:

```
2023-08-18 12:29:45 [Switching to Thread 0x7f79f920d640 (LWP 14744)]
2023-08-18 12:29:45 CRC32Hash::operator() (x=..., this=<optimized out>) at ./base/base/../base/StringRef.h:291
2023-08-18 12:29:45 #0  CRC32Hash::operator() (x=..., this=<optimized out>) at ./base/base/../base/StringRef.h:291
2023-08-18 12:29:45         pos = 0x362e302e37 <error: Cannot access memory at address 0x362e302e37>
2023-08-18 12:29:45         end = 0x46c9318 <vtable for DB::ColumnString+16> ""
2023-08-18 12:29:45         res = <optimized out>
2023-08-18 12:29:45         size = <optimized out>
2023-08-18 12:29:45         word = <optimized out>
2023-08-18 12:29:45         word = <optimized out>
2023-08-18 12:29:45 #1  HashTable<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableNoState>, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::hash (this=<optimized out>, x=...) at ./src/Common/HashTable/HashTable.h:759
2023-08-18 12:29:45 No locals.
2023-08-18 12:29:45 #2  HashTable<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableNoState>, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::find (this=<optimized out>, x=...) at ./src/Common/HashTable/HashTable.h:1116
2023-08-18 12:29:45         place_value = <optimized out>
2023-08-18 12:29:45         hash_value = <optimized out>
2023-08-18 12:29:45 #3  DB::ColumnsHashing::columns_hashing_impl::HashMethodBase<DB::ColumnsHashing::HashMethodString<StringRef, void, true, false, false, false>, StringRef, void, false, false, false>::findKeyImpl<HashSetTable<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, Ha
2023-08-18 12:29:45         it = <optimized out>
2023-08-18 12:29:45         offset = <optimized out>
2023-08-18 12:29:45 #4  DB::ColumnsHashing::columns_hashing_impl::HashMethodBase<DB::ColumnsHashing::HashMethodString<StringRef, void, true, false, false, false>, StringRef, void, false, false, false>::findKey<HashSetTable<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, HashTa
2023-08-18 12:29:45         key_holder = {key = {data = 0x362e302e37 <error: Cannot access memory at address 0x362e302e37>, size = 18446743841080632545}, pool = /build/gdb-ZgDh0V/gdb-12.1/gdb/dwarf2/expr.c:607: internal-error: coerce_pieced_ref: Assertion `closure->pieces.size () == 1\' failed.
2023-08-18 12:29:45 A problem internal to GDB has been detected,
2023-08-18 12:29:45 further debugging may prove unreliable.
```

File: docs/en/operations/settings/merge-tree-settings.md [hunk 1]
```diff
@@ -623,6 +623,19 @@ Possible values:
...
 
+## number_of_free_entries_in_pool_to_execute_optimize_entire_partition {#number_of_free_entries_in_pool_to_execute_optimize_entire_partition}
+
+When there is less than specified number of free entries in pool, do not execute optimizing entire partition in the background (this task generated when set `min_age_to_force_merge_seconds` and enable `min_age_to_force_merge_on_partition_only`). This is to leave free threads for regular merges and avoid "Too many parts".
+
+Possible values:
+
+- Positive integer.
+
+Default value: 20
```
Comment (tavplubix) [inline docs/en/operations/settings/merge-tree-settings.md:634]: It means that background "optimize" has the same priority as mutations. But it should have lower priority, so the default value should be higher

**Review by tavplubix — COMMENTED** [review]
In general LGTM

**Comment by tavplubix** [comment]
> Segmentation fault in tsan:

It's https://github.com/ClickHouse/ClickHouse/issues/52540

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Stress test (debug) - https://github.com/ClickHouse/ClickHouse/issues/37686
Upgrade check (debug) - known issue
Upgrade check (tsan) - maybe https://github.com/ClickHouse/ClickHouse/pull/53799

**Pull request merged**